File: src/msft-recommendation/src/main/python/TrainValidRecommendSplit.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in the project root for information.

import numpy as np
import pyspark
import pyspark.sql.functions as F
import sys
from mmlspark.TrainTestSplit import *
from mmlspark.TrainValidRecommendSplitModel import TrainValidRecommendSplitModel as tvmodel
from pyspark import keyword_only
from pyspark.ml import Estimator
from pyspark.ml.param import Params, Param, TypeConverters
from pyspark.ml.tuning import ValidatorParams
from pyspark.ml.util import *
from pyspark.sql import Window
from pyspark.sql.functions import col, expr

if sys.version >= '3':
    basestring = str


@inherit_doc
class TrainValidRecommendSplit(Estimator, ValidatorParams):
    trainRatio = Param(Params._dummy(), "trainRatio", "Param for ratio between train and\
         validation data. Must be between 0 and 1.", typeConverter=TypeConverters.toFloat)
    userCol = Param(Params._dummy(), "userCol",
                    "userCol: column name for user ids. Ids must be within the integer value range. (default: user)")
    ratingCol = Param(Params._dummy(), "ratingCol", "ratingCol: column name for ratings (default: rating)")

    itemCol = Param(Params._dummy(), "itemCol",
                    "itemCol: column name for item ids. Ids must be within the integer value range. (default: item)")

    def setTrainRatio(self, value):
        """
        Sets the value of :py:attr:`trainRatio`.
        """
        return self._set(trainRatio=value)

    def getTrainRatio(self):
        """
        Gets the value of trainRatio or its default value.
        """
        return self.getOrDefault(self.trainRatio)

    def setItemCol(self, value):
        """

        Args:

            itemCol (str): column name for item ids. Ids must be within the integer value range. (default: item)

        """
        self._set(itemCol=value)
        return self

    def getItemCol(self):
        """

        Returns:

            str: column name for item ids. Ids must be within the integer value range. (default: item)
        """
        return self.getOrDefault(self.itemCol)

    def setRatingCol(self, value):
        """

        Args:

            ratingCol (str): column name for ratings (default: rating)

        """
        self._set(ratingCol=value)
        return self

    def getRatingCol(self):
        """

        Returns:

            str: column name for ratings (default: rating)
        """
        return self.getOrDefault(self.ratingCol)

    def setUserCol(self, value):
        """

        Args:

            userCol (str): column name for user ids. Ids must be within the integer value range. (default: user)

        """
        self._set(userCol=value)
        return self

    def getUserCol(self):
        """

        Returns:

            str: column name for user ids. Ids must be within the integer value range. (default: user)
        """
        return self.getOrDefault(self.userCol)

    @keyword_only
    def __init__(self, estimator=None, estimatorParamMaps=None, evaluator=None, seed=None):
        """
        __init__(self, estimator=None, estimatorParamMaps=None, evaluator=None, numFolds=3,\
                 seed=None)
        """
        super(TrainValidRecommendSplit, self).__init__()
        kwargs = self._input_kwargs
        self._set(**kwargs)

    @keyword_only
    def setParams(self, estimator=None, estimatorParamMaps=None, evaluator=None, seed=None):
        """
        setParams(self, estimator=None, estimatorParamMaps=None, evaluator=None, numFolds=3,\
                  seed=None):
        Sets params for cross validator.
        """
        kwargs = self._input_kwargs
        return self._set(**kwargs)

    def copy(self, extra=None):
        """
        Creates a copy of this instance with a randomly generated uid
        and some extra params. This copies creates a deep copy of
        the embedded paramMap, and copies the embedded and extra parameters over.

        :param extra: Extra parameters to copy to the new instance
        :return: Copy of this instance
        """
        if extra is None:
            extra = dict()
        newCV = Params.copy(self, extra)
        if self.isSet(self.estimator):
            newCV.setEstimator(self.getEstimator().copy(extra))
        # estimatorParamMaps remain the same
        if self.isSet(self.evaluator):
            newCV.setEvaluator(self.getEvaluator().copy(extra))
        return newCV

    def _create_model(self, java_model):
        model = tvmodel()
        model._java_obj = java_model
        model._transfer_params_from_java()
        return model

    def _fit(self, dataset):
        rating = self.getOrDefault(self.ratingCol)
        est = self.getOrDefault(self.estimator)
        eva = self.getOrDefault(self.evaluator)
        epm = self.getOrDefault(self.estimatorParamMaps)
        num_models = len(epm)
        metrics = [0.0] * num_models

        customerID = self.getOrDefault(self.userCol)
        itemID = self.getOrDefault(self.itemCol)

        pyspark.sql.DataFrame.min_rating_filter = TrainTestSplit.min_rating_filter
        pyspark.sql.DataFrame.stratified_split = TrainTestSplit.stratified_split

        temp_train, temp_validation = dataset \
            .dropDuplicates() \
            .withColumnRenamed(self.getUserCol(), 'customerID') \
            .withColumnRenamed(self.getItemCol(), 'itemID') \
            .min_rating_filter(min_rating=6, by_customer=True) \
            .stratified_split(min_rating=3, by_customer=True, fixed_test_sample=False, ratio=0.5)

        train = temp_train \
            .withColumnRenamed('customerID', self.getUserCol()) \
            .withColumnRenamed('itemID', self.getItemCol())

        validation = temp_validation \
            .withColumnRenamed('customerID', self.getUserCol()) \
            .withColumnRenamed('itemID', self.getItemCol())

        train.cache()
        validation.cache()
        # eva.setNItems(validation.rdd.map(lambda r: r[1]).distinct().count())
        print("fit starting")

        numModels = len(epm)

        def prepare_test_data(dataset, recs, k):

            userColumn = est.getUserCol()
            itemColumn = est.getItemCol()

            perUserRecommendedItemsDF = recs \
                .select(userColumn, "recommendations." + itemColumn) \
                .withColumnRenamed(itemColumn, 'prediction')

            windowSpec = Window.partitionBy(userColumn).orderBy(col(rating).desc())

            perUserActualItemsDF = dataset \
                .select(userColumn, itemColumn, rating, F.rank().over(windowSpec).alias('rank')) \
                .where('rank <= {0}'.format(k)) \
                .groupBy(userColumn) \
                .agg(expr('collect_list(' + itemColumn + ') as label')) \
                .select(userColumn, "label")

            joined_rec_actual = perUserRecommendedItemsDF \
                .join(perUserActualItemsDF, on=userColumn) \
                .drop(userColumn)

            return joined_rec_actual

        models = est.fit(train, epm)
        for j in range(numModels):
            model = models[j]  # models[j]
            recs = model.recommendForAllUsers(eva.getK())
            prepared_test = prepare_test_data(model.transform(validation), recs, eva.getK())
            metric = eva.evaluate(prepared_test)
            metrics[j] += metric
        if eva.isLargerBetter():
            bestIndex = np.argmax(metrics)
        else:
            bestIndex = np.argmin(metrics)

        best_model = est.fit(dataset, epm[bestIndex])
        return self._copyValues(tvmodel(best_model, metrics))

    # def getEstimatorParamMaps(self):
    #     # Load information from java_stage to the instance.
    #     estimator = JavaParams._from_java(self.getEstimator())
    #     epms = [estimator._transfer_param_map_from_java(epm)
    #             for epm in self.getEstimatorParamMaps()]
    #
    #     return epms

    # def setEstimatorParamMaps(self, value):
    #     """
    #     Args:
    #         estimatorParamMaps (object): param maps for the estimator
    #
    #     """
    #
    #     gateway = SparkContext._gateway
    #     cls = SparkContext._jvm.org.apache.spark.ml.param.ParamMap
    #
    #     java_epms = gateway.new_array(cls, len(value))
    #
    #     for idx, epm in enumerate(value):
    #         java_epms[idx] = self.getEstimator()._transfer_param_map_to_java(epm)
    #
    #     self._set(estimatorParamMaps=java_epms)
    #     return self

    # def fit(self, dataset):
    #     return self._call_java("fit", dataset)
    #
    # def _fit(self, dataset):
    #     rating = self.getOrDefault(self.ratingCol)
    #     est = self.getOrDefault(self.estimator)
    #     eva = self.getOrDefault(self.evaluator)
    #     epm = self.getOrDefault(self.estimatorParamMaps)
    #     num_models = len(epm)
    #     metrics = [0.0] * num_models
    #
    #     customerID = self.getOrDefault(self.userCol)
    #     itemID = self.getOrDefault(self.itemCol)
    #
    #     def filter_ratings(dataset):
    #         minRatingsU = 4
    #
    #         tmpDF = dataset \
    #             .groupBy(customerID) \
    #             .agg({itemID: "count"}) \
    #             .withColumnRenamed('count(' + itemID + ')', 'nitems') \
    #             .where(col('nitems') >= minRatingsU)
    #
    #         minRatingsI = 4
    #
    #         tmp_df2 = dataset \
    #             .groupBy(itemID) \
    #             .agg({customerID: "count"}) \
    #             .withColumnRenamed('count(' + customerID + ')', 'ncustomers') \
    #             .where(col('ncustomers') >= minRatingsI)
    #
    #         input_df = tmp_df2 \
    #             .join(dataset, itemID) \
    #             .drop('ncustomers') \
    #             .join(tmpDF, customerID) \
    #             .drop('nitems')
    #
    #         return input_df
    #
    #     print("filter dataset")
    #     filtered_dataset = filter_ratings(dataset.dropDuplicates())
    #
    #     def split_df(input_df):
    #         # Stratified sampling by item into a train and test data set
    #         nusers_by_item = input_df \
    #             .groupBy(itemID) \
    #             .agg({customerID: "count"}) \
    #             .withColumnRenamed('count(' + customerID + ')', 'nusers').rdd
    #
    #         perm_indices = nusers_by_item.map(lambda r: (r[0], np.random.permutation(r[1]), r[1]))
    #
    #         RATIO = 0.75
    #         tr_idx = perm_indices.map(lambda r: (r[0], r[1][: int(round(r[2] * RATIO))]))
    #         test_idx = perm_indices.map(lambda r: (r[0], r[1][int(round(r[2] * RATIO)):]))
    #
    #         tr_by_item = input_df.rdd \
    #             .groupBy(lambda r: r[1]) \
    #             .join(tr_idx) \
    #             .flatMap(lambda r: [x for x in r[1][0]][: len(r[1][1])])
    #
    #         test_by_item = input_df.rdd \
    #             .groupBy(lambda r: r[1]) \
    #             .join(test_idx) \
    #             .flatMap(lambda r: [x for x in r[1][0]][: len(r[1][1])])
    #
    #         train_temp = tr_by_item \
    #             .map(lambda r: tuple(r)) \
    #             .toDF(schema=[customerID, itemID, customerID + 'Org', itemID + 'Org', rating])
    #
    #         test_temp = test_by_item \
    #             .map(lambda r: tuple(r)) \
    #             .toDF(schema=[customerID, itemID, customerID + 'Org', itemID + 'Org', rating])
    #
    #         return [train_temp, test_temp]
    #
    #     print("split dataset")
    #     # [train, validation] = self._call_java("splitDF", filtered_dataset)
    #     splitter = TrainTestSplit(dataset.dropDuplicates(), by_customer=True, customerCol=customerID, itemCol=itemID)
    #
    #     train, validation = splitter.stratified_split(3, fixed_test_sample=False, ratio=0.5)
    #     # [train, validation] = split_df(filtered_dataset)
    #     train.cache()
    #     validation.cache()
    #     eva.setNumberItems(validation.rdd.map(lambda r: r[1]).distinct().count())
    #     print("fit starting")
    #     models = est.fit(train, epm)
    #     train.unpersist()
    #
    #     def prepare_test_data(dataset, recs, k):
    #
    #         userColumn = est.getUserCol()
    #         itemColumn = est.getItemCol()
    #
    #         perUserRecommendedItemsDF = recs \
    #             .select(userColumn, "recommendations." + itemColumn) \
    #             .withColumnRenamed(itemColumn, 'prediction')
    #
    #         windowSpec = Window.partitionBy(userColumn).orderBy(col(rating).desc())
    #
    #         perUserActualItemsDF = dataset \
    #             .select(userColumn, itemColumn, rating, F.rank().over(windowSpec).alias('rank')) \
    #             .where('rank <= {0}'.format(k)) \
    #             .groupBy(userColumn) \
    #             .agg(expr('collect_list(' + itemColumn + ') as label')) \
    #             .select(userColumn, "label")
    #
    #         joined_rec_actual = perUserRecommendedItemsDF \
    #             .join(perUserActualItemsDF, on=userColumn) \
    #             .drop(userColumn)
    #
    #         return joined_rec_actual
    #
    #     def cast_model(model):
    #         recs = model._call_java("recommendForAllUsers", eva.getK())
    #         prepared_test = prepare_test_data(model.transform(validation), recs, eva.getK())
    #         metric = eva.evaluate(prepared_test)
    #         return metric
    #
    #     for j in range(num_models):
    #         metrics[j] += cast_model(models[j])
    #     validation.unpersist()
    #
    #     if eva.isLargerBetter():
    #         best_index = np.argmax(metrics)
    #     else:
    #         best_index = np.argmin(metrics)
    #     best_model = est.fit(dataset, epm[best_index])
    #     return self._copyValues(tvmodel(best_model=best_model, validationMetrics=metrics))

File: src/msft-recommendation/src/main/scala/TrainValidRecommendSplit.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.ml.spark

import org.apache.spark.ml._
import org.apache.spark.ml.evaluation.Evaluator
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.recommendation._
import org.apache.spark.ml.util._
import org.apache.spark.sql.expressions.Window
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.{DataFrame, Dataset, Row}

import scala.collection.mutable
import scala.util.Random

@InternalWrapper
class TrainValidRecommendSplit(override val uid: String) extends Estimator[TrainValidRecommendSplitModel]
  with TrainValidRecommendSplitParams with ComplexParamsWritable with MsftRecommendationParams {

  /** @group setParam */
  def setUserCol(value: String): this.type = set(userCol, value)

  /** @group setParam */
  def setItemCol(value: String): this.type = set(itemCol, value)

  /** @group setParam */
  def setRatingCol(value: String): this.type = set(ratingCol, value)

  def this() = this(Identifiable.randomUID("TrainValidRecommendSplit"))

  /** @group setParam */
  def setEstimator(value: Estimator[_ <: Model[_]]): this.type = set(estimator, value)

  /** @group setParam */
  def setEstimatorParamMaps(value: Array[ParamMap]): this.type = set(estimatorParamMaps, value)

  /** @group setParam */
  def setEvaluator(value: Evaluator): this.type = set(evaluator, value)

  /** @group setParam */
  def setTrainRatio(value: Double): this.type = set(trainRatio, value)

  /** @group setParam */
  def setSeed(value: Long): this.type = set(seed, value)

  /** @group setParam */
  def setMinRatingsU(value: Int): this.type = set(minRatingsU, value)

  /** @group setParam */
  def setMinRatingsI(value: Int): this.type = set(minRatingsI, value)

  override def transformSchema(schema: StructType): StructType = transformSchemaImpl(schema)

  override def fit(dataset: Dataset[_]): TrainValidRecommendSplitModel = {
    val schema = dataset.schema
    transformSchema(schema, logging = true)
    val est = $(estimator)
    val eval = $(evaluator).asInstanceOf[MsftRecommendationEvaluator]
    val epm = $(estimatorParamMaps)
    val numModels = epm.length
    val metrics = new Array[Double](epm.length)
    dataset.cache()
    eval.setNItems(dataset.agg(countDistinct(col($(itemCol)))).take(1)(0).getLong(0))
    val filteredDataset = filterRatings(dataset.dropDuplicates())

    val Array(trainingDataset, validationDataset): Array[DataFrame] = splitDF(filteredDataset)
    trainingDataset.cache()
    validationDataset.cache()
    val models = est.fit(trainingDataset, epm).asInstanceOf[Seq[Model[_]]]

    def calculateMetrics(model: Transformer, validationDataset: Dataset[_]): Double = model match {
      case p: PipelineModel => {
        //Assume Rec is last stage of pipeline
        val modelTemp = model.asInstanceOf[PipelineModel].stages.last
        calculateMetrics(modelTemp, validationDataset)
      }
      case m: MsftRecommendationModel => {
        val recs = model.asInstanceOf[MsftRecommendationModel].recommendForAllUsers(5 * eval.getK)
        val preparedTest: Dataset[_] = prepareTestData(model.transform(validationDataset), recs, eval.getK)
        eval.evaluate(preparedTest)
      }
      case a: ALSModel => {
        val recs = model.asInstanceOf[ALSModel].recommendForAllUsers(5 * eval.getK)
        val preparedTest: Dataset[_] = prepareTestData(model.transform(validationDataset), recs, eval.getK)
        eval.evaluate(preparedTest)
      }
      case s: SARModel => {
        val recs = model.asInstanceOf[SARModel].recommendForAllUsers(5 * eval.getK)
        val preparedTest: Dataset[_] = prepareTestData(model.transform(validationDataset), recs, eval.getK)
        eval.evaluate(preparedTest)
      }
      //      case trait: MsftRecommendationModelParams => {
      //        val recs = model.asInstanceOf[MsftRecommendationModelParams].recommendForAllUsers(5 * eval.getK)
      //        val preparedTest: Dataset[_] = prepareTestData(model.transform(validationDataset), recs, eval.getK)
      //        eval.evaluate(preparedTest)
      //      }
    }

    var i = 0
    while (i < numModels) {
      // TODO: duplicate evaluator to take extra params from input
      val metric = calculateMetrics(models(i), validationDataset)
      logDebug(s"Got metric $metric for model trained with ${epm(i)}.")
      metrics(i) += metric
      i += 1
    }

    val (bestMetric, bestIndex) =
      if (eval.isLargerBetter) metrics.zipWithIndex.maxBy(_._1)
      else metrics.zipWithIndex.minBy(_._1)

    val bestModel = est.fit(dataset, epm(bestIndex)).asInstanceOf[Model[_]]
    copyValues(new TrainValidRecommendSplitModel(uid, bestModel, metrics).setParent(this))
  }

  override def copy(extra: ParamMap): TrainValidRecommendSplit = defaultCopy(extra)

  private def filterByItemCount(dataset: Dataset[_]): DataFrame = {
    dataset
      .groupBy($(userCol))
      .agg(col($(userCol)), count(col($(itemCol))))
      .withColumnRenamed("count(" + $(itemCol) + ")", "nitems")
      .where(col("nitems") >= $(minRatingsU))
      .drop("nitems")
      .cache()
  }

  private def filterByUserRatingCount(dataset: Dataset[_]): DataFrame = dataset
    .groupBy($(itemCol))
    .agg(col($(itemCol)), count(col($(userCol))))
    .withColumnRenamed("count(" + $(userCol) + ")", "ncustomers")
    .where(col("ncustomers") >= $(minRatingsI))
    .join(dataset, $(itemCol))
    .drop("ncustomers")
    .cache()

  def filterRatings(dataset: Dataset[_]): DataFrame = filterByUserRatingCount(dataset)
    .join(filterByItemCount(dataset), $(userCol))

  def splitDF(dataset: DataFrame): Array[DataFrame] = {
    val shuffleFlag = true
    val shuffleBC = dataset.sparkSession.sparkContext.broadcast(shuffleFlag)

    val wrapColumn = udf((itemId: Double, rating: Double) => Array(itemId, rating))
    val sliceudf = udf(
      (r: mutable.WrappedArray[Array[Double]]) => r.slice(0, math.round(r.length * $(trainRatio)).toInt))
    val shuffle = udf((r: mutable.WrappedArray[Array[Double]]) =>
      if (shuffleBC.value) Random.shuffle(r.toSeq)
      else r
    )
    val dropudf = udf((r: mutable.WrappedArray[Array[Double]]) => r.drop(math.round(r.length * $(trainRatio)).toInt))
    val popLeft = udf((r: mutable.WrappedArray[Double]) => r(0))
    val popRight = udf((r: mutable.WrappedArray[Double]) => r(1))

    val testds = dataset
      .withColumn("itemIDRating", wrapColumn(col($(itemCol)), col($(ratingCol))))
      .groupBy(col($(userCol)))
      .agg(collect_list(col("itemIDRating")))
      .withColumn("shuffle", shuffle(col("collect_list(itemIDRating)")))
      .withColumn("train", sliceudf(col("shuffle")))
      .withColumn("test", dropudf(col("shuffle")))
      //      .withColumn("train2", explode(sliceudf(col("shuffle"))))
      //      .withColumn($(itemCol) + "train2", popLeft(col("train2")))
      //      .withColumn($(ratingCol) + "train2", popRight(col("train2")))
      //      .withColumn("test2", explode(dropudf(col("shuffle"))))
      //      .withColumn($(itemCol) + "test2", popLeft(col("test2")))
      //      .withColumn($(ratingCol) + "test2", popRight(col("test2")))
      //      .withColumn("train", sliceudf(col("collect_list(itemIDRating)")))
      //      .withColumn("test", dropudf(col("collect_list(itemIDRating)")))
      .drop(col("collect_list(itemIDRating)")).drop(col("shuffle"))
      .cache()

    //    val train2 = testds.select(
    //      col("customerID"),
    //      col($(itemCol) + "train2").as($(itemCol)),
    //      col($(ratingCol) + "train2").as($(ratingCol))
    //    )
    //    val test2 = testds.select(
    //      col("customerID"),
    //      col($(itemCol) + "test2").as($(itemCol)),
    //      col($(ratingCol) + "test2").as($(ratingCol))
    //    )

    val train = testds.select($(userCol), "train")
      .withColumn("itemIdRating", explode(col("train")))
      .drop("train")
      .withColumn($(itemCol), popLeft(col("itemIdRating")))
      .withColumn($(ratingCol), popRight(col("itemIdRating")))
      .drop("itemIdRating")

    val test = testds.select($(userCol), "test")
      .withColumn("itemIdRating", explode(col("test")))
      .drop("test")
      .withColumn($(itemCol), popLeft(col("itemIdRating")))
      .withColumn($(ratingCol), popRight(col("itemIdRating")))
      .drop("itemIdRating")

    Array(train, test)
  }

  def prepareTestData(validationDataset: DataFrame, recs: DataFrame, k: Int): Dataset[_] = {
    import org.apache.spark.sql.functions.{collect_list, rank => r}

    val est = $(estimator) match {
      case p: Pipeline => {
        //Assume Rec is last stage of pipeline
        val pipe = $(estimator).asInstanceOf[Pipeline].getStages.last
        pipe match {
          case m: MsftRecommendation => {
            pipe.asInstanceOf[MsftRecommendation]
          }
          case a: ALS => {
            pipe.asInstanceOf[ALS]
          }
          case s: SAR => {
            pipe.asInstanceOf[SAR]
          }
        }
      }
      case m: MsftRecommendation => {
        $(estimator).asInstanceOf[MsftRecommendation]
      }
      case a: ALS => {
        $(estimator).asInstanceOf[ALS]
      }
      case s: SAR => {
        $(estimator).asInstanceOf[SAR]
      }
    }

    val userColumn = est.getUserCol
    val itemColumn = est.getItemCol

    val perUserRecommendedItemsDF: DataFrame = recs
      .select(userColumn, "recommendations." + itemColumn)
      .withColumnRenamed(itemColumn, "prediction")

    val windowSpec = Window.partitionBy(userColumn).orderBy(col($(ratingCol)).desc)

    val perUserActualItemsDF = validationDataset
      .select(userColumn, itemColumn, $(ratingCol))
      .withColumn("rank", r().over(windowSpec).alias("rank"))
      .where(col("rank") <= k)
      .groupBy(userColumn)
      .agg(col(userColumn), collect_list(col(itemColumn)))
      .withColumnRenamed("collect_list(" + itemColumn + ")", "label")
      .select(userColumn, "label")

    val joined_rec_actual = perUserRecommendedItemsDF
      .join(perUserActualItemsDF, userColumn)
      .drop(userColumn)

    joined_rec_actual
  }
}

object TrainValidRecommendSplit extends ComplexParamsReadable[TrainValidRecommendSplit] {
  def popRow(r: Row): Any = r.getDouble(1)
}
